torch/tensortypes/nestedtensor/nested.py

```diff
@@ -0,0 +1,215 @@
+import torch
+
+# This implementation is based on NestedTensor 0.0.1
+# RFC: https://github.com/pytorch/pytorch/issues/22169
+
+def is_nested_tensor(obj):
+    return isinstance(obj, NestedTensor)
+
+
+# Arguments match torch.tensor
+def make_nested_tensor(data, dtype=None, device=None, requires_grad=False, pin_memory=False):
+    if is_nested_tensor(data):
+        # This is consistent with torch.tensor(torch.Tensor)
+        # but errors out.
+        raise ValueError("UserWarning: To copy construct from a NestedTensor, "
+                         "use sourceTensor.clone().detach() or "
+                         "sourceTensor.clone().detach().requires_grad_(True), "
+                         "rather than torch.tensor(sourceTensor).")
+    elif torch.is_tensor(data):
+        # The user has the right to expect a NestedTensor from this
+        # function, but we can't meaningfully provide one if passed a Tensor
+        raise ValueError("Can't construct a NestedTensor from a Tensor")
+    else:
+        if not (isinstance(data, list) or isinstance(data, tuple)):
+            raise ValueError("Pass a list or tuple to construct NestedTensor.")
+        for data_ in data:
+            if not torch.is_tensor(data_):
+                raise ValueError("Each element of the tuple or list must "
+                                 "be a torch.Tensor")
+        tensors = []
+        for data_ in data:
+            # torch.tensor copies on construction
+            new_data = torch.empty_like(data_)
+            new_data.copy_(data_)
+            new_data = new_data.to(dtype)
+            new_data = new_data.to(device)
+            new_data = new_data.requires_grad_(requires_grad)
+            if pin_memory:
+                new_data = new_data.pin_memory()
+            tensors.append(new_data)
+
+        return NestedTensor(tensors)
+
+def as_nestedtensor(data, dtype=None, device=None):
+    ret = NestedTensor(data)
+    if dtype is not None:
+        ret = ret.to(dtype)
+    if device is not None:
+        ret = ret.to(device)
+    return ret
+
+class NestedTensor():
+    # The attributes must match across all constiuents
+    # and default to the empty Tensor's if the given list
+    # is empty.
+    #
+    # The NestedTensor's attributes then become that of its
+    # constiuents.
+    #
+    # Attributes:
+    #     dim
+    #     layout
+    #     device
+    #     dtype
+    #     requires_grad
+    #     is_pinned
+    def __init__(self, tensors):
+        for tensor in tensors:
+            assert torch.is_tensor(tensor)
+        self.tensors = tensors
```

Comment:
Doesn't feel safe for `self.tensors` to be a public API

Reply:
But it's not. It's only used within internal operators.

Reply:
I guess I can add an underscore to make that more explicit.